File: src/main/java/icu/samnyan/aqua/sega/maimai2/handler/impl/UploadUserPlaylogHandler.java
package icu.samnyan.aqua.sega.maimai2.handler.impl;

import com.fasterxml.jackson.core.JsonProcessingException;

import icu.samnyan.aqua.sega.maimai2.dao.userdata.UserDataRepository;
import icu.samnyan.aqua.sega.maimai2.dao.userdata.UserPlaylogRepository;
import icu.samnyan.aqua.sega.maimai2.handler.BaseHandler;
import icu.samnyan.aqua.sega.maimai2.model.request.UploadUserPlaylog;
import icu.samnyan.aqua.sega.maimai2.model.userdata.UserDetail;
import icu.samnyan.aqua.sega.maimai2.model.userdata.UserPlaylog;
import icu.samnyan.aqua.sega.util.jackson.BasicMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

import java.util.Map;
import java.util.Optional;

/**
 * @author samnyan (privateamusement@protonmail.com)
 */
@Component("Maimai2UploadUserPlaylogHandler")
public class UploadUserPlaylogHandler implements BaseHandler {

    private static final Logger logger = LoggerFactory.getLogger(UploadUserPlaylogHandler.class);

    private final BasicMapper mapper;
    private final UserPlaylogRepository userPlaylogRepository;
    private final UserDataRepository userDataRepository;

    public UploadUserPlaylogHandler(UserDataRepository userDataRepository, UserPlaylogRepository userPlaylogRepository, BasicMapper mapper) {
        this.userDataRepository = userDataRepository;
        this.userPlaylogRepository = userPlaylogRepository;
        this.mapper = mapper;
    }

    @Override
    public String handle(Map<String, Object> request) throws JsonProcessingException {
        UploadUserPlaylog uploadUserPlaylog = mapper.convert(request, UploadUserPlaylog.class);

        Optional<UserDetail> userOptional = userDataRepository.findByCardExtId(uploadUserPlaylog.getUserId());

        /*
        Due to how we handle userId, first user playlog can't be saved.
        (sequence order swapped, it sends playlog then user detail)
        It might be possible to fix this with some workaround, but leave it like this at this time.
        */
        if (userOptional.isPresent()) {
            UserDetail userDetail = userOptional.get();

            UserPlaylog userPlaylog = uploadUserPlaylog.getUserPlaylog();

            userPlaylog.setUser(userDetail);
            userPlaylogRepository.save(userPlaylog);
        }

        return "{\"returnCode\":1,\"apiName\":\"com.sega.maimai2servlet.api.UploadUserPlaylogApi\"}";
    }
}

File: src/main/java/icu/samnyan/aqua/sega/maimai2/handler/UploadUserPlaylogHandler.kt
package icu.samnyan.aqua.sega.maimai2.handler

import ext.millis
import icu.samnyan.aqua.sega.maimai2.model.Mai2UserDataRepo
import icu.samnyan.aqua.sega.maimai2.model.Mai2UserPlaylogRepo
import icu.samnyan.aqua.sega.general.BaseHandler
import icu.samnyan.aqua.sega.maimai2.model.request.UploadUserPlaylog
import icu.samnyan.aqua.sega.maimai2.model.userdata.UserPlaylog
import icu.samnyan.aqua.sega.util.jackson.BasicMapper
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Component
import kotlin.jvm.optionals.getOrNull

/**
 * @author samnyan (privateamusement@protonmail.com)
 */
@Component("Maimai2UploadUserPlaylogHandler")
class UploadUserPlaylogHandler(
    private val userDataRepository: Mai2UserDataRepo,
    private val playlogRepo: Mai2UserPlaylogRepo,
    private val mapper: BasicMapper
) : BaseHandler {
    data class BacklogEntry(val time: Long, val playlog: UserPlaylog)
    companion object {
        @JvmStatic
        val playBacklog = mutableMapOf<Long, MutableList<BacklogEntry>>()
    }

    override fun handle(request: Map<String, Any>): String {
        val req = mapper.convert(request, UploadUserPlaylog::class.java)

        // Save if the user is registered
        val u = userDataRepository.findByCardExtId(req.userId).getOrNull()
        if (u != null) playlogRepo.save(req.userPlaylog.apply { user = u })

        // If the user hasn't registered (first play), save the playlog to a backlog
        else {
            playBacklog.putIfAbsent(req.userId, mutableListOf())
            playBacklog[req.userId]?.apply {
                add(BacklogEntry(millis(), req.userPlaylog))
                if (size > 6) clear()  // Prevent abuse
            }
        }

        return """{"returnCode":1,"apiName":"com.sega.maimai2servlet.api.UploadUserPlaylogApi"}"""
    }

    @Scheduled(fixedDelay = 60_000)
    fun cleanBacklog() {
        // Clean all backlog entries that are older than 5 minutes
        val now = millis()
        playBacklog.filter { (_, v) -> v.isEmpty() || v[0].time - now > 300_000 }.toList()
            .forEach { (k, _) -> playBacklog.remove(k) }
    }
}
